### PR TITLE
Sema: Solve Bind constraints involving to two variables of different lvalue-ness [5.0]

### PR DESCRIPTION
**Pull request by slavapestov**
The restriction here only applies to Equal constraints.

Fixes <rdar://problem/45511838>, <https://bugs.swift.org/browse/SR-9068>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@xedin Are you OK with pulling this into 5.0? The code in CSSimplify has been like that forever but for some reason the test case in the radar regressed from 4.2 to 5.0. It's a user project so we should fix it IMO, but its your call since I know CSSimplify is very subtle in general.

**Comment by slavapestov** [comment]
@swift-ci Please test

File: lib/Sema/CSSimplify.cpp
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: @xedin Note the old behavior here was kind of weird and used the word 'materializable' that I don't like:

- CanBindToLValue meant can bind to LValueType *or* InOutType
- CanBindToInOut meant can bind to InOutType

I don't think this was intended. I fixed it but I don't think there's a behavior change in practice since the only time we see InOutType in type variables is with BindParam constraint.

File: test/Constraints/generics.swift
Comment (xedin) [inline test/Constraints/generics.swift]: Nit: newline is missing :)

**Review by xedin — APPROVED** [review]
I think this is low risk and we should take it, especially because it fixes on of the source compatibility issues.

File: test/Constraints/generics.swift
Comment (xedin) [inline test/Constraints/generics.swift]: This is very nice!

**Comment by xedin** [comment]
As a follow-up could you create PR to un-XFAIL `NonEmpty` project? I can do that too if you'd like

**Comment by slavapestov** [comment]
I’ll do it once @AnnaZaks merges this PR

**Review by AnnaZaks — APPROVED** [review]

**Pull request merged**
